x/gasless/keeper/fee_helper.go

```diff
@@ -0,0 +1,241 @@
+		if consumption.GasTankId == gasTank.Id {
+			consumptionIndex = index
+			consumptionFound = true
+		}
+	}
+	// no need to check the consumption usage since there is no key available with given gas tank id
+	// i.e the consumer has never used this gas reserve before and the first time visitor for the given gas tank
+	if !consumptionFound {
+		return gasTank, true, nil
+	}
+
+	consumptionDetails := consumer.Consumptions[consumptionIndex]
+
+	// consumer is blocked by the gas tank
+	if consumptionDetails.IsBlocked {
+		return gasTank, false, sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "blocked by gas tank")
+	}
+
+	// consumer exhausted the transaction count limit, hence not eligible with given gas tank
+	if consumptionDetails.TotalTxsMade >= consumptionDetails.TotalTxsAllowed {
+		return gasTank, false, sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "exhausted tx limit")
+	}
+
+	// if total fees consumed by the consumer is more than or equal to the allowed consumption
+	// i.e consumer has exhausted its fee limit and hence is not eligible for the given tank
+	totalFeeConsumption := consumptionDetails.TotalFeesConsumed.Add(fee.Amount)
+	if totalFeeConsumption.GTE(consumptionDetails.TotalFeeConsumptionAllowed) {
+		return gasTank, false, sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "exhausted total fee usage or pending fee limit insufficient for tx")
+	}
+
+	return gasTank, true, nil
+}
+
+func (k Keeper) GetFeeSource(ctx sdk.Context, sdkTx sdk.Tx, originalFeePayer sdk.AccAddress, fees sdk.Coins) sdk.AccAddress {
+	if len(sdkTx.GetMsgs()) > 1 {
+		k.EmitFeeConsumptionEvent(ctx, originalFeePayer, []uint64{}, []error{sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "multiple messages")}, 0)
+		return originalFeePayer
+	}
+
+	// only one fee coin is supported, tx containing multiple coins as fees are not allowed.
+	if len(fees) != 1 {
+		k.EmitFeeConsumptionEvent(ctx, originalFeePayer, []uint64{}, []error{sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "asked fee != 1")}, 0)
+		return originalFeePayer
+	}
+
+	fee := fees[0]
+
+	msg := sdkTx.GetMsgs()[0]
+	msgTypeURL := sdk.MsgTypeURL(msg)
+
+	isContract := false
+	var contractAddress string
+
+	executeContractMessage, ok := msg.(*wasmtypes.MsgExecuteContract)
+	if ok {
+		isContract = true
+		contractAddress = executeContractMessage.GetContract()
+	}
+
+	txIdentifier := msgTypeURL
+	if isContract {
+		txIdentifier = contractAddress
+	}
+
+	// check if there are any gas tansk for given txIdentifier i.e msgTypeURL or Contract address
+	// if there is no gas tank for the given identifier, fee source will be original feePayer
+	txGtids, found := k.GetTxGTIDs(ctx, txIdentifier)
+	if !found {
+		k.EmitFeeConsumptionEvent(ctx, originalFeePayer, []uint64{}, []error{sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "no gas tanks found")}, 0)
+		return originalFeePayer
+	}
+
+	tempConsumer, found := k.GetGasConsumer(ctx, originalFeePayer)
+	if !found {
```

Comment:
when can !found happen, please comment

Reply:
the address which have never consumed a fee from any of the gas tank, can return not found